pkg/ha/ha_service.go

```diff
@@ -0,0 +1,142 @@
+package ha
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"golang.org/x/sync/semaphore"
+
+	"github.com/timescale/promscale/pkg/log"
+	"github.com/timescale/promscale/pkg/pgxconn"
+)
+
+const (
+	backOffDurationOnLeaderChange = 10 * time.Second
+	haSyncerTimeInterval          = 15 * time.Second
+)
+
+type Service struct {
+	state             map[string]*State
+	lockClient        haLockClient
+	leaseTimeout      time.Duration
+	leaseRefresh      time.Duration
+	_leaderChangeLock *semaphore.Weighted
+	_syncTicker       *time.Ticker
+}
+
+func NewHAService(dbClient pgxconn.PgxConn) (*Service, error) {
+	lockClient := newHaLockClient(dbClient)
+	timeout, refresh, err := lockClient.readLeaseSettings(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("could not create HA state: %#v", err)
+	}
+
+	service := &Service{
+		state:             make(map[string]*State),
+		lockClient:        lockClient,
+		leaseTimeout:      timeout,
+		leaseRefresh:      refresh,
+		_leaderChangeLock: semaphore.NewWeighted(1),
+		_syncTicker:       time.NewTicker(haSyncerTimeInterval),
+	}
+	go service.haStateSyncer()
+	return service, nil
+}
+
+func (h *Service) haStateSyncer() {
+	for range h._syncTicker.C {
+		for clusterName, state := range h.state {
+			// To validate cluster state we use maxTimeInstance instead if state.leader
+			// because if prometheus leader crashes promscale needs to sync leader
+			// using maxTimeSeenInstance and validate if leaderHasChanged condition
+			err := h.validateClusterState(state.leaseStart, state.maxTimeSeen, clusterName, state.maxTimeInstance)
+			if err != nil {
+				log.Error("failed to validate cluster %s state %v", clusterName, err)
+			}
+		}
+	}
+}
+
+// checkInsert verifies the samples are from prom leader & in expected time range.
+// 	returns true if the insert is allowed
+func (h *Service) checkInsert(minT, maxT time.Time, clusterName, replicaName string) (bool, error) {
```

Comment:
I would suggest calling this as `areSamplesFromLeader` which would make the func self-explanatory and returning true would mean **yes, it is from leader**